### PR TITLE
TOOLING-255: Finalizing resque pause

### PR DESCRIPTION
**Pull request by wedy**
awaitng for
https://github.com/chrisboulton/php-resque/pull/213
and https://github.com/chrisboulton/php-resque/pull/212

once merged, this will good to go

### CODE REVIEW

**Comment by gabelimon** [comment]
Alright, this thing has seen a major facelift. Tests are now passing and I've added PSR-4 compliance. I did a bit of renaming which is open to bikeshedding.
@bigcommerce/tools @wedy @aleachjr @chrisboulton @lord2800

File: .travis.yml [hunk 1]
```diff
@@ -12,3 +12,6 @@ before_script:
   - sh -c "if [ $REDIS_STANDALONE -eq 0 ]; then cd phpredis-2.2.3/ && phpize && ./configure && make && make install; fi"
   - sh -c "if [ $REDIS_STANDALONE -eq 0 ]; then echo \"extension=redis.so\" >> `php --ini | grep \"Loaded Configuration\" | sed -e \"s|.*:\s*||\"`; fi"
   - composer install
+script:
+  - ./vendor/bin/phpcs --standard=PSR2 src/ test/
```
Comment (lord2800) [inline .travis.yml]: Might be useful to wrap this up in a [composer script](https://getcomposer.org/doc/articles/scripts.md).

File: composer.json [hunk 2]
```diff
@@ -1,25 +1,26 @@
...
+  "repositories": [{
+    "type": "git",
+    "url": "https://github.com/chrisboulton/php-resque"
+  }],
+  "require": {
+    "chrisboulton/php-resque": "dev-master"
+  },
   "require-dev": {
-    "phpunit/phpunit": "3.7.*"
+    "phpunit/phpunit": "3.7.*",
```
Comment (lord2800) [inline composer.json]: Unless you have a particular reason to lock to this old version of phpunit, the latest stable version is 4.5.

File: composer.json [hunk 3]
```diff
@@ -1,25 +1,26 @@
...
+    "squizlabs/php_codesniffer": "*"
```
Comment (lord2800) [inline composer.json]: You should lock this to a specific version (2.3.0 is the latest).

File: composer.json (hunk 2 above)
Comment (gabelimon) [inline composer.json]: Thanks.

File: .travis.yml (hunk 1 above)
Comment (gabelimon) [inline .travis.yml]: Will do

File: phpunit.xml [hunk 4]
```diff
@@ -12,14 +12,14 @@
...
+			<directory>./test/</directory>
 		</testsuite>
 	</testsuites>
 
 	<filter>
 		<whitelist>
-			<directory suffix=".php">./lib/ResquePause/</directory>
-			<directory suffix=".php">./lib/</directory>
+			<directory suffix=".php">./src/</directory>
+            <directory suffix=".php">./src/Jobs/</directory>
```
Comment (lord2800) [inline phpunit.xml]: Whitespace is different here.

File: src/Pause.php [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+ * Base Resque Pause class
+ *
+ * @package        PHP Resque Pause
+ * @author         Wedy Chainy <askwedi@wedipedia.org>
+ * @licence        http://www.opensource.org/licenses/mit-license.php
+ */
+class Pause
+{
+    /**
+     * Pause the job create a flag, and rename the original queue to temporary queue
```
Comment (lord2800) [inline src/Pause.php]: This is unclear grammatically.

File: composer.json (hunk 3 above)
Comment (gabelimon) [inline composer.json]: Thanks

File: phpunit.xml (hunk 4 above)
Comment (gabelimon) [inline phpunit.xml]: Looks like I left in some tabs.

File: test/Jobs/PauseTest.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+ * @license     http://www.opensource.org/licenses/mit-license.php
+ */
+class JobTest extends PHPUnit_Framework_TestCase
+{
+    protected $resque;
+    protected $redis;
+
+    public function setUp()
+    {
+        $config = file_get_contents(REDIS_CONF);
```
Comment (lord2800) [inline test/Jobs/PauseTest.php]: Is there a reason you're testing against a real, live redis instance instead of mocking out redis entirely?

Comment (gabelimon) [inline test/Jobs/PauseTest.php]: I assume it's because that's how the php-resque tests work but I'm all for mocking it

File: test/PauseTest.php [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        Resque::enqueue('upgrade:test8', 'test');
+        Resque::enqueue('upgrade:test8', 'test');
+
+        # pause it!
+        Pause::pause('upgrade:test8');
+
+        try {
+            #  before enqueue pauseCallback
+            Pause::pauseCallback('test', array(), 'upgrade:test8', 'dummyid');
+        } catch (Resque_Job_DontCreate $e) {
```
Comment (lord2800) [inline test/PauseTest.php]: Could you give a bit more explanation as to why you're catching this? I'd like to see an assert (above, in the try) that describes why it's OK to ignore this.

File: test/PauseTest.php [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    {
+        Resque::enqueue('upgrade:test7', 'test');
+        Pause::pause('upgrade:test7');
+        Pause::unpause('upgrade:test7');
+        $this->assertFalse((bool)Pause::isPaused('upgrade:test7'));
+    }
+
+    public function testPauseCallback()
+    {
+        # we have 2 in queue:upgrade:test8
```
Comment (lord2800) [inline test/PauseTest.php]: Please use `//` for comments.

File: test/PauseTest.php (hunk 7 above)
Comment (gabelimon) [inline test/PauseTest.php]: Ah, I actually meant to change this behavior. Sorry about that.

File: test/PauseTest.php (hunk 8 above)
Comment (gabelimon) [inline test/PauseTest.php]: I'm surprised the codesniffer didn't catch this.

File: test/bootstrap.php [hunk 9]
```diff
@@ -15,134 +16,24 @@
...
+    exit(1);
 
 }
 
 // Get redis port from conf
 $config = file_get_contents(REDIS_CONF);
-if(!preg_match('#^\s*port\s+([0-9]+)#m', $config, $matches)) {
-	echo "Could not determine redis port from redis.conf";
-	exit(1);
+if (!preg_match('#^\s*port\s+([0-9]+)#m', $config, $matches)) {
```
Comment (lord2800) [inline test/bootstrap.php]: I've seen this same thing done 3 different times now--maybe it'd be good to make this a function?

File: test/PauseTest.php (hunk 8 above)
Comment (lord2800) [inline test/PauseTest.php]: Me too.

File: test/bootstrap.php (hunk 9 above)
Comment (gabelimon) [inline test/bootstrap.php]: If the tests don't use redis we don't need this file at all.

File: test/Jobs/PauseTest.php (hunk 6 above)
Comment (gabelimon) [inline test/Jobs/PauseTest.php]: So it turns out resque really doesn't allow for very good dependency injection. Redis server is set by this function meaning that unless I have a mock redis instance sitting on a port I'm not going to be able to mock this.
```
	public static function setBackend($server, $database = 0)
	{
		self::$redisServer   = $server;
		self::$redisDatabase = $database;
		self::$redis         = null;
	}```

File: LICENSE [hunk 10]
```diff
@@ -1,4 +1,5 @@
-(c) 2014 Wedy Chainy <askwedi@wedipedia.org>
+Copyright (C) BigCommerce, 2014
```
Comment (lord2800) [inline LICENSE]: Nitpick, but I believe the official name is `Bigcommerce` these days. You should probably check with someone more knowledgeable than me though.

**Comment by gabelimon** [comment]
The README needs to be updated to fit the new interface but I figured I'd leave this for last.

File: src/Pause.php [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    public function __construct()
+    {
+        // Create object for pausing jobs
+        $this->pauser = new JobPauser(Resque::redis(), \Resque_Redis::getPrefix());
+        $pause = $this;
+        $pauser = $this->pauser;
+
+        // Listen for enqueue and move any new jobs to temp queue
+        $this->listener = function ($class, $args, $queue, $id) use (&$pause, $pauser) {
+            if ($pause->isPaused($queue)) {
```
Comment (gabelimon) [inline src/Pause.php]: This is the one place where `$this` is used here. I'm considering switching it for the underlying `$pauser` function which had to be passed in for the `pushPausedJob` method anyway. Opinions?

Comment (lord2800) [inline src/Pause.php]: Seems reasonable.

File: phpunit.xml [hunk 12]
```diff
@@ -1,25 +1,25 @@
...
-	backupStaticAttributes="false"
-	colors="true"
-	convertErrorsToExceptions="true"
-	convertNoticesToExceptions="true"
-	convertWarningsToExceptions="true"
-	processIsolation="false"
-	stopOnFailure="false"
-	syntaxCheck="false"
+    backupStaticAttributes="false"
+    bootstrap="Tests/bootstrap.php"
```
Comment (lord2800) [inline phpunit.xml]: Change this to point phpunit straight at the composer `autoload.php` file. One less file that serves no purpose other than to include another file.

Comment (gabelimon) [inline phpunit.xml]: Ah, you're smart.

**Comment by gabelimon** [comment]
@lord2800 are you fairly happy with this now?
Also I'd appreciate other people reviewing as well guise.

File: Tests/Integration/PauseTest.php [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    public function testIsPaused()
+    {
+        Resque::enqueue('upgrade:test5', 'test');
+        $this->pauser->pause('upgrade:test5');
+        $this->assertTrue($this->pauser->isPaused('upgrade:test5'));
+    }
+
+    public function testIsPaused2()
```
Comment (lord2800) [inline Tests/Integration/PauseTest.php]: Could you come up with better names for these two functions?

**Comment by lord2800** [comment]
Seems good to me other than that.

File: Tests/Integration/PauseTest.php (hunk 13 above)
Comment (gabelimon) [inline Tests/Integration/PauseTest.php]: The options in my mind are
1) Give them really long names
2) Mash these three tests into a single test

Comment (lord2800) [inline Tests/Integration/PauseTest.php]: Really long names > `testIsPaused{,2,3}`. Do these tests make sense to each be in isolation?

Comment (gabelimon) [inline Tests/Integration/PauseTest.php]: I agree entirely. There's not technical reason to separate them.

**Comment by gabelimon** [comment]
Compounded some of the tests. Now they test multiple behaviors of the same method.

File: Tests/Integration/PauseTest.php [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+        $this->assertTrue($this->pauser->isPaused('upgrade:test5'));
+
+        // Non-paused queue
+        Resque::enqueue('upgrade:test6', 'test');
+        $this->assertFalse($this->pauser->isPaused('upgrade:test6'));
+
+        // Paused and resumed queue
+        Resque::enqueue('upgrade:test7', 'test');
+        $this->pauser->pause('upgrade:test7');
+        $this->pauser->resume('upgrade:test7');
```
Comment (lord2800) [inline Tests/Integration/PauseTest.php]: Not critical, but might be worth an assert to verify that the queue is paused after calling pause. Might not be important in the end, but I generally prefer over-cautious use of asserts than not.

**Comment by lord2800** [comment]
I can keep nitpicking for eternity, but :+1: at this point.

File: README.md [hunk 15]
```diff
@@ -7,8 +7,51 @@ resque-pause adds functionality to pause resque jobs
...
+```
+
+### Resume it! ###
+```php
+$GLOBALS['ResquePause']->resume('My_Queue');
+```
+
+### Is it Paused? ###
+```php
+$GLOBALS['ResquePause']e->isPaused('My_Queue');
```
Comment (wedy) [inline README.md]: minor, just a typo ``e`` there

**Comment by wedy** [comment]
LGTM @gabelimon  :+1:

File: README.md (hunk 15 above)
Comment (gabelimon) [inline README.md]: Good catch.

**Comment by cwalsh** [comment]
:shipit:

**Pull request merged**